### DISCUSSION TITLE
Circular dependency between WiFi and lwIP-Ethernet ?

### DISCUSSION THREAD

**Discussion by ruifig** [earlephilhower/arduino-pico#1408] [General]
I'm using PlatformIO for my project, and since it's a bit of a complex project, I'm now at the point the `lib_ldf_mode=chain+`  I'm using takes ages to process. 
I'm now trying to set `lib_ldf_mode` to `off`  and specifying all the libraries manually.

I'm almost there (I think), but seems like I came a cross a circular dependency.

My Dependency graph (according to PlatformIO) is this:
```
Dependency Graph
|-- TFT_eSPI @ 2.5.30
|   |-- SPI @ 1.0
|   |-- LittleFS @ 0.1.0
|-- czmicromuc @ 0.0.1
|-- ArduinoMqtt @ 1.5.1+sha.15091f0
|-- Adafruit_MCP23017
|   |-- Wire @ 1.0
|-- AT24C
|   |-- Wire @ 1.0
|   |-- czmicromuc @ 0.0.1
|-- Wire @ 1.0
|-- WiFi @ 1.0.0     <<<<<<<<<<<<<<<<<<<<  THIS
|   |-- Updater @ 1.0
|   |   |-- MD5Builder @ 1.0.0
|   |   |-- LittleFS @ 0.1.0
|   |   |-- PicoOTA @ 1.0.0
|   |   |   |-- LittleFS @ 0.1.0
|   |-- MD5Builder @ 1.0.0
|   |-- lwIP-Ethernet @ 1     <<<<<<<<<<<<<<<<<<<< THIS
|   |   |-- lwIP_CYW43 @ 1
|   |   |   |-- SPI @ 1.0
|   |   |-- SPI @ 1.0
|   |-- lwIP_CYW43 @ 1
|   |   |-- SPI @ 1.0
|   |-- SPI @ 1.0
|-- Updater @ 1.0
|   |-- MD5Builder @ 1.0.0
|   |-- LittleFS @ 0.1.0
|   |-- PicoOTA @ 1.0.0
|   |   |-- LittleFS @ 0.1.0
|-- MD5Builder @ 1.0.0
|-- SPI @ 1.0
|-- LittleFS @ 0.1.0
|-- XPT2046_Touchscreen
|   |-- SPI @ 1.0
|   |-- czmicromuc @ 0.0.1
|-- Adafruit HTU21DF Library @ 1.0.5
|   |-- Wire @ 1.0
```

`LwipEthernet.cpp` needs a header from the `WiFi` library. This:

```
In file included from C:\Users\Rui\.platformio\packages\framework-arduinopico\libraries\lwIP_CYW43\src/lwIP_CYW43.h:3,
                 from C:\Users\Rui\.platformio\packages\framework-arduinopico\libraries\lwIP_Ethernet\src/LwipEthernet.h:28,
                 from C:\Users\Rui\.platformio\packages\framework-arduinopico\libraries\lwIP_Ethernet\src\LwipEthernet.cpp:2:
C:\Users\Rui\.platformio\packages\framework-arduinopico\libraries\lwIP_Ethernet\src/LwipIntfDev.h:48:10: fatal error: wl_definitions.h: No such file or directory
```

If let LFD on, it seems PlatformIO solves it by just adding the right `-I` parameters, so the circular dependency is not a problem.
But when specifying the libraries manually that doesn't work because the `-I` parameters are added according to the library order.
For example, given these libs:
```
lib_deps =
    A
    B
    C
```

C depends on B, and B depends on A.
When compiling A, A won't see the headers for B or C. 

While writing this post I actually found a solution. I can just add a `-I` for WiFi to `build_flags`
This:
`-I "${platformio.packages_dir}/framework-arduinopico/libraries/WiFi/src"`

Still, starting this discussion because maybe it's something that can be fixed at the Arduino-Pico level for the future?

**Comment by earlephilhower** [comment]
@ruifig can you undo the hacked extra `-I` and give #1415 a try and report back?

**Reply by ruifig** [reply]
Sure. I'll let you know. Probably later today or tomorrow.

**Comment by ruifig** [comment]
Your suggested fix gets past that error, but then another one appears. :)

WiFi and Updater libraries also have a circular dependency, because of StackThunk.h/cpp
Once I moved StackThunk.h and StackThunk.cpp to cores/rp2040 then it built fine.

**Reply by earlephilhower** [reply]
I'll update the PR later tonight and we should be all set. Thanks for the report!
